src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs

```diff
@@ -275,14 +272,24 @@ public bool MoveNext()
                         Current = next;
                         return true;
                     }
-                    else if (++_segmentIndex >= _tokenizer.Count)
+                    // Next template has the same length as the url we are trying to match
+                    // The only possible matching segments are either our current matches or
+                    // any catch-all segment after this segment in which the catch all is empty.
+                    else if (next.Length == _tokenizer.Count)
```

Comment:
Discussed in person - the issue here was that we weren't correctly updating `_segementIndex` on all code paths. As @javiercn has discovered, we don't really need `_segmentIndex` we can just look at the current node's Length.